craft_parts/parts.py

```diff
@@ -0,0 +1,243 @@
+# -*- Mode:Python; indent-tabs-mode:nil; tab-width:4 -*-
+#
+# Copyright 2021 Canonical Ltd.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Definitions and helpers to handle parts."""
+
+import copy
+from pathlib import Path
+from typing import Any, Dict, List, Optional, Set
+
+from craft_parts import errors
+from craft_parts.dirs import ProjectDirs
+from craft_parts.steps import Step
+
+
+class Part:
+    """Each of the components used in the project specification.
+
+    During the craft-parts lifecycle each part is processed through
+    different steps in order to obtain its final artifacts. The Part
+    class holds the part specification data and additional configuration
+    information used during step processing.
+
+    :param name: The part name.
+    :param data: A dictionary containing the part properties.
+    :param project_dirs: The project work directories.
+    """
+
+    def __init__(
+        self,
+        name: str,
+        data: Dict[str, Any],
+        *,
+        project_dirs: ProjectDirs = None,
```

Comment:
I'm curious why project_dirs is required here (and not just part_dir). Can you open a draft PR of the working tree so I can see an up-to-date big picture? pretty plz :D

Reply:
We'll have a set of `ProjectInfo`/`PartInfo`/`StepInfo` classes that are used to pass information to callbacks. `ProjectInfo` is created first and instantiates `ProjectDirs` from a given `work_dir`, then passes it to each `Part` instance so that the path of each project directory is defined only once -- if later we decided that each dir should be specified independently this would only affect `ProjectDirs`. If we create `Part` based only on `parts_dir` we would also need to redefine the paths in the `Part` initialization. The context will appear in the next PR which is currently being prepared, I'll provide a draft soon.

Reply:
OK thanks